### PR TITLE
New glue builder

### PR DESCRIPTION
**Pull request by PaulWessel**
**Description of proposed changes**

See #3180 and #3130 for content.  This PR addresses type 3 supplements that are built long after GMT has been built and installed.  To enable these supplements to have their modules listed in the `gmt --help` listing these maintainers only need to

1. Run `gmt --new-glue=name > gmt_name_suppl.c` in their source directory, with name the desired name of their shared library (e.g., mbsystem).
2. Include gmt_name_suppl.c when building the shared library.

Note: Unlike the automatic building of these glues via CMake configuration, the above command only creates a single source code where the *moduleinfo* part is inlined directly and not a separately configured file.  The modules are automatically sorted alphabetically.

### CODE REVIEW

**Comment by PaulWessel** [comment]
PS. The only thing I don't like about this implementation is the fact that I am hardwiring in lots of C code based on gmt_glue.c.in.  We will have to remember to update two places.  I could not think of a simple way to read the gmt_glue.c.in at run time.  I could write a build_glue.sh script that creates an include snippet for gmt_support.c like we did in the old gmt_module_*.c|h days.  Comments welcome.

**Comment by seisman** [comment]
A quick question: Can type 3 supplements access low-level APIs? Is this repo https://github.com/GenericMappingTools/custom-supplements type 3?

**Comment by PaulWessel** [comment]
I will need to have a look at that given the recent cnanges.  This may be a type 4 supplement which is only relying on gmt.h.  if so they cannot make use of any glue (since depends on gmt_internals.h).

**Comment by PaulWessel** [comment]
Finally, this call does not look for or handle a **SHARED_LIB_PURPOSE** setting.  I could look for an optional

`#define THIS_MODULE_LIB_PURPOSE "MB-System: GMT compatible modules"`

string in the modules and pick the first I find, else it just gets the name of the library.
Any better ideas?

**Comment by seisman** [comment]
> This PR addresses type 3 supplements that are built long after GMT has been built and installed.

I'm confused why type 3 supplements like mbsystem can access low-level APIs, since GMT is already built and installed.

**Comment by PaulWessel** [comment]
The mbsystem is a type 2 supplement because in has

`#include "gmt_dev.h"`

and that means it has access to the gmt_* and gmtlib_* functions of the GMT library.  Those supplements (I guess type 4) that only include gmt.h are limited to the GMT_* functions.

For many packages, there is a separate "-dev" install, e.g., gmt-dev, that would install the gmt_dev.h and other include files, while the main GMT package would only install gmt.h (and a few more called by gmt.h).

So may need to review the type numbering

1. Official GMT supplment in src/*
2. Wannabee supplement elsewhere but bilt with GMT via newsuppl symbolic link to external src
3. Outside supplement (e.g. mbsystem) that includes gmt_dev.h and thus can use glue to connet to gmt
4. External supplements only relying on the API (gmt.h).  No glue possible.

**Comment by PaulWessel** [comment]
I think my suggestion above is not so bad for external type 3 supplements.  In at least one of the module C codes, include the line (e.g.)

`#define THIS_MODULE_LIB_PURPOSE "MB-System: GMT compatible modules"
`
to set the shared library purpose string.  Any better ideas?

**Comment by PaulWessel** [comment]
And if no such thing is found I can print a warning to suggest they edit that line in the glue file.

**Comment by seisman** [comment]
> I think my suggestion above is not so bad for external type 3 supplements. In at least one of the module C codes, include the line (e.g.)
> 
> `#define THIS_MODULE_LIB_PURPOSE "MB-System: GMT compatible modules" `
> to set the shared library purpose string. Any better ideas?

I don't have a better solution for this.

File: doc/rst/source/gmt.rst [hunk 1]
```diff
@@ -79,6 +79,13 @@ If no module is given then several other options are available:
     scripting language among *bash*, *csh*, or *batch*.  Default is the main shell
     closest to your current shell (e.g., bash for zsh, csh for tcsh).
 
+**--new-glue**\ =\ *name*
+    Write the C code glue needed when building third-party supplements as shared
+    libraries.  The *name* is the name of the shared library. Run **gmt** in the directory
+    of the supplement and the glue code will be written to *stdout*.  Including this C code
+    when building the shared library means **gmt** can list available modules via the
+    **-show-modules**, **--help** options.  We recommend saving the code to gmt_*name*_glue.c.
```
Comment (seisman) [inline doc/rst/source/gmt.rst:87]:
```suggestion
    **--show-modules**, **--help** options.  We recommend saving the code to gmt\_\ *name*\_glue.c.
```

File: src/gmt_support.c [hunk 2]
```diff
@@ -16659,3 +16659,141 @@ char * gmt_add_options (struct GMT_CTRL *GMT, const char *list) {
...
+		error = GMT_RUNTIME_ERROR;
+		goto CROAK;
+	}
+
+	n++;
+	GMT_Report (API, GMT_MSG_INFORMATION, "%d module files found in current directory\n", n);
+
+	if (first_purpose) {
+		GMT_Report (API, GMT_MSG_NOTICE, "No #define THIS_MODULE_LIB_PURPOSE setting found in any module.  Please edit argument to gmtlib_module_show_all\n");
+		lib_purpose = library;
```
Comment (seisman) [inline src/gmt_support.c:16740]: Change the default library purpose to?
```
GMT xxx: The third-party supplements to the Generic Mapping Tools
```

Comment (PaulWessel) [reply src/gmt_support.c]: Done

**Comment by PaulWessel** [comment]
testgmtio.c had a THIS_MODULE_OPTIONS that triggered my test...

**Pull request merged**